File: AddTwoNumbers.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x, next=None):
        self.val = x
        self.next = next


class Solution:
    # @return a ListNode
    def addTwoNumbers(self, l1, l2):
            """Advance l1 and l2 together, aware carry
            Time: O(n)
            Space: O(1)
            Corner: empty list, carry > 0
            """
            curr1, curr2 = l1, l2
            curr = dummy = ListNode(0)
            num = 0
            while curr1 or curr2:
                num /= 10  # as carry
                if curr1:
                    num += curr1.val
                    curr1 = curr1.next
                if curr2:
                    num += curr2.val
                    curr2 = curr2.next
                curr.next = ListNode(num % 10)
                curr = curr.next

            if num / 10 == 1:  # max
                curr.next = ListNode(1)
            return dummy.next

if __name__ == "__main__":
    l1 = ListNode(9, ListNode(9, ListNode(9)))
    l2 = ListNode(1)
    l = Solution().addTwoNumbers(l1, l2)
    while l:
        print l.val
        l = l.next

File: problems/add-two-numbers/add_two_numbers.rb
# Definition for singly-linked list.
class ListNode
  attr_accessor :val, :next
  def initialize(val)
      @val = val
      @next = nil
  end
end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
  dummy = ListNode.new(nil)
  current = dummy
  carry = 0
  while l1 || l2 || carry != 0
    sum = carry + (l1 ? l1.val : 0) + (l2 ? l2.val : 0)
    current.next = ListNode.new(sum % 10)
    carry = sum / 10

    current = current.next
    l1 = l1 ? l1.next : nil
    l2 = l2 ? l2.next : nil
  end
  dummy.next
end

def print_list(l)
  while l
    p l.val
    l = l.next
  end
end
l1 = ListNode.new(10)
l2 = ListNode.new(9)
print_list(add_two_numbers(l1, l2))
